Add Etcd compatibility test

liggitt opened the pull request:
/kind cleanup

#### What this PR does / why we need it:

Builds on https://github.com/kubernetes/kubernetes/pull/99871, only the last commit is new

Fixes https://github.com/kubernetes/kubernetes/issues/99863

#### Special notes for your reviewer:

For every version we iterate over and check storage versions for, inspect the API compatibility data from previous releases.

If a previous release knew about a non-alpha version of the group-kind, then expect the current storage version to be understood by a previous release.

Would have caught https://github.com/kubernetes/kubernetes/pull/99662#discussion_r588075892 with this error:

>`    --- FAIL: TestEtcdStoragePath/discovery.k8s.io/v1,_Resource=endpointslices (0.10s)`
>`        etcd_storage_path_test.go:161: Previous releases understand non-alpha versions ["v1beta1"], but do not understand the expected current storage version "v1". This means a current server will store data in etcd that is not understood by a previous version.`
>`    --- FAIL: TestEtcdStoragePath/discovery.k8s.io/v1beta1,_Resource=endpointslices (0.03s)`
>`        etcd_storage_path_test.go:161: Previous releases understand non-alpha versions ["v1beta1"], but do not understand the expected current storage version "v1". This means a current server will store data in etcd that is not understood by a previous version.`


```release-note
NONE
```

liggitt committed:
Ensure storage version is understood by prior releases

fedebongio commented:
/assign @wenjiaswe @ptabor 
/triage accepted

wenjiaswe commented:
Thanks for doing this!

ptabor commented:
/assign @caesarxuchao as this is domain of storage-version-migrator

caesarxuchao commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -139,6 +140,31 @@ func TestEtcdStoragePath(t *testing.T) {
+					expectedGVK.Version,
+				)
+			}
+
```
This block of change lgtm.

liggitt commented:
rebased, added positive check that we could locate the compatibility fixtures

test-only change, important for API compatibility checks

/milestone v1.21

liggitt commented:
/retest

wenjiaswe commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -181,8 +235,10 @@ func TestEtcdStoragePath(t *testing.T) {
+
 func dumpEtcdKVOnFailure(t *testing.T, kvClient clientv3.KV) {
-	if t.Failed() {
+	if t.Failed() && debug {
```
Is debug setting to true anywhere?

liggitt replied:
no. this output is excessively verbose and is not relevant for most test failures... keeping this here is fine, but requiring someone to flip the debug var to true when debugging a failure locally

wenjiaswe replied:
Thanks for the explanation.

wenjiaswe commented:
/lgtm
/approve
